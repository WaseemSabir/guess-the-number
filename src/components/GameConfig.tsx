import React from 'react';

interface GameConfigProps {
    resetNumber: () => void,
    lowerBound: number,
    setLowerBound: (new_val: number) => void,
    upperBound: number,
    setUpperBound: (new_val: number) => void,
}

function GameConfig(props: GameConfigProps) {
    const {lowerBound, upperBound, resetNumber, setUpperBound, setLowerBound} = props

    return (
        <>
            <h3>Game Config</h3>
            <div>
                <label data-testid="lower-bound-label">Lower bound: </label>
                <input
                    data-testid="lower-bound-input"
                    id="lower-bound-input"
                    type='number'
                    value={lowerBound}
                    onChange={e => setLowerBound(Number(e.target.value))}
                />
            </div>
            <div>
                <label data-testid="upper-bound-label">Upper bound: </label>
                <input
                    data-testid="upper-bound-input"
                    id="upper-bound-input"
                    type='number'
                    value={upperBound}
                    onChange={e => setUpperBound(Number(e.target.value))}
                />
            </div>
            <button onClick={resetNumber} data-testid="reset-button">Reset</button>
        </>
    );
}

export default GameConfig;
