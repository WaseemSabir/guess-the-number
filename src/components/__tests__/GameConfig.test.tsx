import {cleanup, fireEvent, render} from "@testing-library/react";
import React from "react";
import GameConfig from "../GameConfig";
import {screen} from '@testing-library/dom';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from "react-test-renderer";


afterEach(cleanup)
Enzyme.configure({adapter: new Adapter()});

it('should render without crashes', () => {
    const {unmount} = render(getGameConfig());
    unmount();
});

it("should trigger reset on click", () => {
    const baseProps = {
        resetNumber: jest.fn()
    }
    render(getGameConfig(baseProps));
    fireEvent.click(screen.getByTestId("reset-button"));
    expect(baseProps.resetNumber).toHaveBeenCalledTimes(1);
});

it("Should trigger set lower bound when value changes", () => {
    const baseProps = {
        setLowerBound: jest.fn()
    }
    const wrapper = shallow(getGameConfig(baseProps));
    wrapper.find("input").at(0).simulate('change', {target: {value: 2}});
    expect(baseProps.setLowerBound).toHaveBeenCalledTimes(1);
    expect(baseProps.setLowerBound).toBeCalledWith(2);
});

it("Should trigger set upper bound when value changes", () => {
    const baseProps = {
        setUpperBound: jest.fn()
    }
    const wrapper = shallow(getGameConfig(baseProps));
    wrapper.find("input").at(1).simulate('change', {target: {value: 100}});
    expect(baseProps.setUpperBound).toHaveBeenCalledTimes(1);
    expect(baseProps.setUpperBound).toBeCalledWith(100);
});

it('Match snapshot', () => {
    const component = renderer.create(
        getGameConfig()
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})

function getGameConfig(
    props: any = {}
) {
    let {resetNumber, setLowerBound, setUpperBound, upperBound, lowerBound} = props

    if (!resetNumber) {
        resetNumber = () => {
        }
    }

    if (!setLowerBound) {
        setLowerBound = (v: number) => {
        }
    }
    if (!setUpperBound) {
        setUpperBound = (v: number) => {
        }
    }
    if (!upperBound) {
        upperBound = 10
    }
    if (!lowerBound) {
        lowerBound = 1
    }

    return (
        <GameConfig
            resetNumber={resetNumber}
            setLowerBound={setLowerBound}
            setUpperBound={setUpperBound}
            upperBound={upperBound}
            lowerBound={lowerBound}
        />
    );
}
