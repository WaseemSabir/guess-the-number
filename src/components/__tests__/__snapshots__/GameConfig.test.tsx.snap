// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
Array [
  <h3>
    Game Config
  </h3>,
  <div>
    <label
      data-testid="lower-bound-label"
    >
      Lower bound: 
    </label>
    <input
      data-testid="lower-bound-input"
      id="lower-bound-input"
      onChange={[Function]}
      type="number"
      value={1}
    />
  </div>,
  <div>
    <label
      data-testid="upper-bound-label"
    >
      Upper bound: 
    </label>
    <input
      data-testid="upper-bound-input"
      id="upper-bound-input"
      onChange={[Function]}
      type="number"
      value={10}
    />
  </div>,
  <button
    data-testid="reset-button"
    onClick={[Function]}
  >
    Reset
  </button>,
]
`;
